src/pages/Signup.jsx

```diff
@@ -0,0 +1,154 @@
+import { Link } from "react-router-dom";
+import { TextInput } from "../components";
+import { useDocumentTitle, useAuthForm, useToggle } from "../hooks";
+import { signup } from "../services";
+import { useDispatch } from "react-redux";
+import {
+  EMAIL_REGEX,
+  EMAIL_ACTION,
+  USERNAME_ACTION,
+  PASSWORD_ACTION,
+  CONFIRM_PASSWORD_ACTION,
+  FULL_NAME_ACTION,
+} from "../Utils/constants";
+import { RiLoaderFill } from "../Utils/icons";
+
+function Signup() {
+  useDocumentTitle("Signup");
+  const {
+    authFormState: { fullName, username, email, password, confirmPassword },
+    authFormDispatch,
+  } = useAuthForm();
+
+  const dispatch = useDispatch();
+
+  const { toggle: isLoading, setToggle: setIsLoading } = useToggle(false);
+
+  const signupHandler = async (e) => {
+    e.preventDefault();
+    if (!EMAIL_REGEX.test(email)) {
+      alert("Invalid email address");
+      return;
+    }
+
+    if (
```

Comment:
You can extract out the logic into a separate file.

Reply:
Yes, done!